src/script/descriptor.cpp

```diff
@@ -275,29 +285,49 @@ class BIP32PubkeyProvider final : public PubkeyProvider
     bool IsRange() const override { return m_derive != DeriveType::NO; }
     size_t GetSize() const override { return 33; }
-    bool GetPubKey(int pos, const SigningProvider& arg, CPubKey* key, KeyOriginInfo& info) const override
+    bool GetPubKey(int pos, const SigningProvider& arg, CPubKey* key, KeyOriginInfo& info, const DescriptorCache* read_cache = nullptr, DescriptorCache* write_cache = nullptr) const override
     {
+        KeyOriginInfo info_out, parent_info;
+        CKeyID keyid = m_extkey.pubkey.GetID();
+        std::copy(keyid.begin(), keyid.begin() + sizeof(info_out.fingerprint), info_out.fingerprint);
+        info_out.path = m_path;
+        parent_info = info_out;
+        if (m_derive == DeriveType::UNHARDENED) info_out.path.push_back((uint32_t)pos);
+        if (m_derive == DeriveType::HARDENED) info_out.path.push_back(((uint32_t)pos) | 0x80000000L);
         if (key) {
-            if (IsHardened()) {
-                CKey priv_key;
-                if (!GetPrivKey(pos, arg, priv_key)) return false;
-                *key = priv_key.GetPubKey();
+            CExtPubKey extkey = m_extkey;
+            CExtPubKey parent = extkey;
+            if (read_cache) {
+                if (!read_cache->GetCachedExtPubKey(info_out, extkey)) {
+                    if (m_derive == DeriveType::HARDENED) return false;
+                    // Try to get the derivation parent
+                    if (!read_cache->GetCachedExtPubKey(parent_info, parent)) return false;
+                    assert(m_derive == DeriveType::UNHARDENED);
+                    parent.Derive(extkey, pos);
+                }
+            } else if (IsHardened()) {
+                CExtKey xprv;
+                if (!GetDerivedExtKey(arg, xprv)) return false;
+                parent = xprv.Neuter();
+                if (m_derive == DeriveType::UNHARDENED) xprv.Derive(xprv, pos);
+                if (m_derive == DeriveType::HARDENED) xprv.Derive(xprv, pos | 0x80000000UL);
+                extkey = xprv.Neuter();
             } else {
-                // TODO: optimize by caching
-                CExtPubKey extkey = m_extkey;
                 for (auto entry : m_path) {
-                    extkey.Derive(extkey, entry);
+                    parent.Derive(parent, entry);
                 }
-                if (m_derive == DeriveType::UNHARDENED) extkey.Derive(extkey, pos);
+                extkey = parent;
+                if (m_derive == DeriveType::UNHARDENED) parent.Derive(extkey, pos);
                 assert(m_derive != DeriveType::HARDENED);
-                *key = extkey.pubkey;
+            }
+            *key = extkey.pubkey;
+            if (write_cache && !read_cache) {
+                if (!write_cache->CacheExtPubKey(info_out, extkey)) return false;
+                // Only cache parent if there is any unhardened derivation
+                if (m_derive == DeriveType::UNHARDENED && !write_cache->CacheExtPubKey(parent_info, parent)) return false;
```

Comment:
Is it worth skipping the parent `write` once it's done?

Reply:
A lookup is just as expensive as an insertion, so I don't think so.